src/processor/operator/recursive_extend/recursive_join.cpp

```diff
@@ -137,37 +139,137 @@ bool RecursiveJoin::getNextTuplesInternal(ExecutionContext* context) {
+        if (vectors->pathVector != nullptr) {
+            vectors->pathVector->resetAuxiliaryBuffer();
+        }
+        frontiersScanner->scan(
+            vectors.get(), offsetVectorSize, nodeIDDataVectorSize, relIDDataVectorSize);
+        if (offsetVectorSize == 0) {
+            return false;
+        }
+        vectors->dstNodeIDVector->state->initOriginalAndSelectedSize(offsetVectorSize);
+        return true;
+    } else {
+        while (true) {
+            auto tableID = *begin(dataInfo->dstNodeTableIDs);
+            auto ret = morselDispatcher->writeDstNodeIDAndPathLength(
+                sharedState->inputFTableSharedState, vectorsToScan, colIndicesToScan,
+                vectors->dstNodeIDVector, vectors->pathLengthVector, tableID, bfsMorsel);
+            /**
+             * ret > 0: non-zero path lengths were written to vector, return to parent op
+             * ret < 0: path length writing is complete, proceed to computeBFS for another morsel
+             * ret = 0: the distance morsel received was empty, go back to get another morsel
+             */
+            if (ret > 0) {
+                return true;
+            } else if (ret == 0) {
+                continue;
+            } else {
+                return false;
+            }
+        }
     }
-    frontiersScanner->scan(
-        vectors.get(), offsetVectorSize, nodeIDDataVectorSize, relIDDataVectorSize);
-    if (offsetVectorSize == 0) {
-        return false;
+}
+
+/**
+ * The main computeBFS to be called for both 1T1S and nTkS scheduling policies. Initially for both
+ * cases, the SSSPSharedState ptr will be null.
+ * - For nTkS policy, this will be filled with a pointer to the main SSSPSharedState from which work
+ *   will be fetched. This SSSPSharedState will not be bound to a specific thread and can change.
+ * - For 1T1S policy, there is no SSSPSharedState involved and a thread works on its own morsel, not
+ *   shared with any other thread.
+ */
+bool RecursiveJoin::computeBFS(kuzu::processor::ExecutionContext* context) {
+    if (morselDispatcher->getSchedulerType() == SchedulerType::OneThreadOneMorsel) {
+        std::pair<GlobalSSSPState, SSSPLocalState> state{COMPLETE, NO_WORK_TO_SHARE};
+        auto checkState = morselDispatcher->getBFSMorsel(sharedState->inputFTableSharedState,
+            vectorsToScan, colIndicesToScan, vectors->srcNodeIDVector, bfsMorsel, state);
+        if (checkState) {
+            return false;
+        }
+        computeBFSOneThreadOneMorsel(context);
+        return true;
+    } else {
+        while (true) {
+            if (bfsMorsel->hasSSSPSharedState()) {
+                SSSPLocalState state{NO_WORK_TO_SHARE};
+                bool checkState = bfsMorsel->ssspSharedState->getBFSMorsel(bfsMorsel, state);
+                if (!checkState) {
+                    computeBFSnThreadkMorsel(context);
+                    if (bfsMorsel->ssspSharedState->finishBFSMorsel(bfsMorsel)) {
```

Comment:
this is a bad call chain. Either do `ssspSharedState.finishBFSMorsel(morsel)` or `morsel.finish()`

Reply:
changed